src/sync/subscribable.hpp

```diff
@@ -0,0 +1,110 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_SUBSCRIBABLE_HPP
+#define REALM_SUBSCRIBABLE_HPP
+
+#include <cstdint>
+#include <unordered_map>
+
+namespace realm {
+
+/// Generic subscribable that allows for coarse, manual notifications
+/// from class type T.
+template <class T>
+struct Subscribable {
+    /// Token that identifies an observer.
+    /// Unsubscribes when deconstructed to
+    /// avoid dangling observers.
+    struct Token {
+        Token(Subscribable* subscribable, uint64_t token)
+        : m_subscribable(subscribable)
+        , m_token(token)
+        {
+        }
+        Token(Token&& other)
+        : m_subscribable(std::move(other.m_subscribable))
+        , m_token(std::move(other.m_token))
+        {
+            other.m_subscribable = nullptr;
+        }
+        Token& operator=(Token&& other) = default;
+        Token(const Token&) = delete;
+        Token& operator=(const Token&) = delete;
+
+        ~Token()
+        {
+            if (m_subscribable) {
+                m_subscribable->unsubscribe(*this);
+            }
+        }
+
+        uint64_t value() const
+        {
+            return m_token;
+        }
+    private:
+        Subscribable* m_subscribable;
+        uint64_t m_token;
+
+        template <class U> friend struct Subscribable;
+    };
+
+    using Observer = std::function<void(const T&)>;
+
+    /// Subscribe to notifications for class type T. Any mutation to the T class
+    /// will trigger the observer. Notifying subscribers must be done manually
+    /// by the Subscribable.
+    /// @param observer callback to be called on mutation
+    /// @returns a token identifying the observer
+    [[nodiscard]] Token subscribe(Observer&& observer)
+    {
+        static uint64_t m_token;
+        m_subscribers.insert({m_token, std::move(observer)});
```

Comment:
This type probably needs a mutex which guards access to m_subscribers? App and User are thread-safe so the subscribe functions need to be usable from any thread as well.